Add TimestampType and packing/unpacking support

randallpittman opened the pull request:
Here's my attempt at addressing #316. I made a TimestampType that can be packed/unpacked. I did not add any `datetime` conversions, but it will accept a floating-point seconds value for creation and can output to float with `to_float_s`. Also included a test. 

I tried to think it through pretty well and keep it 2/3 compatible. Let me know what you think.

jfolz commented on msgpack/_packer.pyx:
```diff
@@ -253,6 +254,15 @@ cdef class Packer(object):
+            elif type(o) is TimestampType if strict_types else isinstance(o, TimestampType):
+                # This should also be before Tuple because TimestampType is namedtuple.
+                code = -1
+                data = o.to_bytes()
```
This calls lots of Python functions, so it's probably quite slow.

randallpittman commented:
Ok, sure, I wondered about that. I'll try to get that done in Cython.
Randy
________________________________
From: jfolz <notifications@github.com>
Sent: Saturday, November 23, 2019 12:53:10 AM
To: msgpack/msgpack-python <msgpack-python@noreply.github.com>
Cc: Randy Pittman <randallpittman@outlook.com>; Author <author@noreply.github.com>
Subject: Re: [msgpack/msgpack-python] Add TimestampType and packing/unpacking support (#375)


@jfolz commented on this pull request.

________________________________

In msgpack/_packer.pyx<https://github.com/msgpack/msgpack-python/pull/375#discussion_r349864642>:

> @@ -253,6 +254,15 @@ cdef class Packer(object):
                     raise ValueError("EXT data is too large")
                 ret = msgpack_pack_ext(&self.pk, longval, L)
                 ret = msgpack_pack_raw_body(&self.pk, rawval, L)
+            elif type(o) is TimestampType if strict_types else isinstance(o, TimestampType):
+                # This should also be before Tuple because TimestampType is namedtuple.
+                code = -1
+                data = o.to_bytes()


This calls lots of Python functions, so it's probably quite slow.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/msgpack/msgpack-python/pull/375?email_source=notifications&email_token=ABWJOYOOCVSMT27556L46M3QVDVPNA5CNFSM4JQXANM2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCMX5F3Y#pullrequestreview-321901295>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABWJOYIOXDB7FSSN22JBF6TQVDVPNANCNFSM4JQXANMQ>.

methane commented on msgpack/__init__.py:
```diff
@@ -3,7 +3,12 @@
 from .exceptions import *
 
 from collections import namedtuple
+from numbers import Number, Integral
```
Don't use them.  Just use int.

randallpittman replied:
May I ask why not? As a user I find it obnoxious to have to convert NumPy types to pure Python types and the `Number` and `Integral` general types alleviate that.

methane replied:
> May I ask why not? 

### Consistency

We don't use them in normal integers.  Why do we support it in Timestamp?
It doesn't make sense at all.

### Performance

```
$ python3 -m timeit -s 'from numbers import Integral' -- 'isinstance(42, Integral)'
500000 loops, best of 5: 443 nsec per loop
$ python3 -m timeit 'isinstance(42, int)'
5000000 loops, best of 5: 59.9 nsec per loop
```

---

>  As a user I find it obnoxious to have to convert NumPy types to pure Python types

Do you really want to use numpy types in Timestamp????

randallpittman replied:
Thanks, everything you said makes sense.

methane commented on msgpack/__init__.py:
```diff
@@ -3,7 +3,12 @@
+import struct
+import sys
+if sys.version_info[0] > 2:
+    long = int
```
I will drop Python 2 support before adding timestamp support.
So no need to care about Python 2 compatibility.

methane commented on msgpack/__init__.py:
```diff
@@ -12,10 +17,74 @@ def __new__(cls, code, data):
             raise ValueError("code must be 0~127")
         return super(ExtType, cls).__new__(cls, code, data)
 
+class TimestampType(namedtuple('TimestampType', 'seconds, nanoseconds')):
```
Don't use namedtuple.  Just use class with `__slots__`.

methane commented on msgpack/_packer.pyx:
```diff
@@ -253,6 +254,15 @@ cdef class Packer(object):
                     raise ValueError("EXT data is too large")
                 ret = msgpack_pack_ext(&self.pk, longval, L)
                 ret = msgpack_pack_raw_body(&self.pk, rawval, L)
+            elif type(o) is TimestampType if strict_types else isinstance(o, TimestampType):
```
No need to `isinstance`.  strict_types is just a backward compatibility option.
No need to care about supporting subclass of Timestamp.

randallpittman commented:
OK, I think I've addressed all the comments:
- I've added C packing and unpacking implementations (basically pulled from msgpack-c),
- Removed the Python 2 compat stuff (but still check for python3 in the test so it doesn't fail the CI.)
- TimestampType just expects basic python numeric types
- TimestampType is a simple class with `__slots__`.
- Added docs

Happy to get your thoughts.
-Randy

methane commented:
Thank you, @randallpittman.  I merged this pull request with some modification.

randallpittman commented:
Glad to contribute.
